Add EIP-5568: Required Action Signals Using Revert Reasons (#5568)
ethereum/EIPs: EIPS/eip-5568.md

---
eip: 5568

> title: Revert Signals

Author: How about:
```suggestion
title: Required Action Signaling Through Revert Reasons
```

description: Signal to wallets that an action is needed by returning a custom revert code
author: Pandapip1 (@Pandapip1)
discussions-to: https://ethereum-magicians.org/t/eip-5568-revert-signals/10622
status: Draft
type: Standards Track
category: Interface
created: 2022-08-31
requires: 140
---

## Abstract

> This EIP introduces a minimalistic machine-readable (binary) format to signal to wallets that an action needs to be taken by the user using a well-known revert reason. This custom revert reason contains just enough data to be extendable by future EIPs and to take in arbitrary parameters (up to 64 kB of data). Example use cases could include approving a token for an exchange, sending an HTTP request, or requesting the user to rotate their keys after a certain period of time to enforce good hygene.

Reviewer: ```suggestion
This EIP introduces a minimalistic machine-readable (binary) format to signal to wallets that an action needs to be taken by the user using a well-known revert reason. This custom revert reason contains just enough data to be extendable by future EIPs and to take in arbitrary parameters (up to 64 kB of data). Example use cases could include approving a token for an exchange, sending an HTTP request, or requesting the user to rotate their keys after a certain period of time to enforce good hygiene.
```

### Signal Response

Before submitting a transaction to the mempool, it MUST be evaluated in a local fork of the chain. If it reverts and has a revert reason starting with the bytes `0x15C0`, then it MUST be treated as a signal.

To parse the signal, discard the first 2 bytes (the `0x15C0` magic number). Then, repeat the following until the signal is fully read:

Read the next 3 bytes to determine the `INSTRUCTION_ID`, and the next 2 bytes to determine the number of bytes to read. Then, read that many bytes to get the `INSTRUCTION_DATA`. Finally, evaluate the instruction as defined in its specification.

After all the instructions have been followed, the wallet SHOULD NOT re-evaluate the transaction, except if an instruction explicitly states that the transaction should be re-evaluated.

## Rationale

This EIP was explicitly optimized for deployment gas cost. It is expected that libraries will eventually be developed that makes sending and receiving signals more developer-friendly.

## Backwards Compatibility

Needs discussion.

## Security Considerations

> Needs discussion.

Author: IMO that would be more important for future extensions. I don't think this EIP should try to anticipate what other EIPs will or will not do.

> Needs discussion.

Reviewer: That's very true. You could at least mention that it might be possible to collide with non-signal revert messages? Or explain how it isn't possible. Whichever :P

## Motivation

Oftentimes, a smart contract needs to signal to a wallet that an action needs to be taken, such as to sign a transaction or send an HTTP request to a URL. Traditionally, this has been done by hard-coding the logic into the frontend, but this EIP allows the smart contract itself to request the action.

This means that, for example, an exchange or a market can directly tell the wallet to approve the smart contract to spend the token, vastly simplifying the front-end code.

## Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

### Custom Revert Reason

To send a signal to a wallet, a compliant smart contract MUST revert with the following error:

```solidity

> error WalletSignal24(uint8 number_hint, uint24 instruction_id, bytes instruction_data)

Reviewer 2: I am not sure if `error` is a proper solidity here. I wonder if there is a typo here.

Author: `error` is proper solidity as of 0.8.4: https://blog.soliditylang.org/2021/04/21/custom-errors/

### Signal Response

Before submitting a transaction to the mempool, it MUST be evaluated locally. If it reverts and the revert signature matches the custom error, then it MUST be treated as a signal. (It is RECOMMENDED for wallets to show a warning if the transaction reverts, even if the revert is not a signal).

The `number_hint`, `instruction_id`, and `instruction_data` MUST be parsed from the revert data. It is RECOMMENDED for wallets to show a progress indicator using the `number_hint`. The instruction SHOULD be evaluated as per the relevant EIP. If the instruction is not supported by the wallet, it MUST display an error to the user indicating that is the case. The wallet MUST then re-evaluate the transaction, except if an instruction explicitly states that the transaction MUST NOT be re-evaluated.

If an instruction is invalid, or the `number_hint`, `instruction_id`, and `instruction_data` cannot be parsed, then an error MUST be displayed to the user indicating that is the case.

## Rationale

> This EIP was explicitly optimized for deployment gas cost and simplicity. It is expected that libraries will eventually be developed that makes= sending and receiving signals more developer-friendly.

Reviewer 3: ```suggestion
This EIP was explicitly optimized for deployment gas cost and simplicity. It is expected that libraries will eventually be developed that makes sending and receiving signals more developer-friendly.
```

## Motivation

Oftentimes, a smart contract needs to signal to a wallet that an action needs to be taken, such as to sign a transaction or send an HTTP request to a URL. Traditionally, this has been done by hard-coding the logic into the frontend, but this EIP allows the smart contract itself to request the action.

This means that, for example, an exchange or a market can directly tell the wallet to approve the smart contract to spend the token, vastly simplifying the front-end code.

## Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

### Custom Revert Reason

> To send a signal to a wallet, a smart contract MUST revert with the following [EIP-140](./eip-140.md) revert data:

Reviewer 2: I love this proposal. Minor part is not making it MUST but just RECOMMENDED. Including a reason can be expensive for smart contract.

> To send a signal to a wallet, a smart contract MUST revert with the following [EIP-140](./eip-140.md) revert data:

Reviewer 2: ```suggestion
To send a signal to a wallet, a smart contract is RECOMMENDED to revert with the following [EIP-140](./eip-140.md) revert data:
```
I love this proposal. Minor part is not making it MUST but just RECOMMENDED. Including a reason can be expensive for smart contract.

Author: "Compliant" is implicit, since this is not a core EIP. I will reluctantly add it.
